### PR TITLE
Document MsTeams access plugin

### LINKED ISSUE #15854: Add a guide to the Microsoft Teams Access Request plugin
## Details

This is the only Access Request plugin we don't have a guide for yet. We [merged the PR yesterday](https://github.com/gravitational/teleport-plugins/pull/617), so let's get this out ASAP.

### Category
- New Feature

### PR DESCRIPTION
**Pull request by hugoShaka**
Fixes #15854 and part of #11818 

This PR introduces the guide for MsTeams access request plugin.

Blockers:
- ~the install script is missing, blocked by https://github.com/gravitational/teleport-plugins/pull/638~
- ~posting to channels is not a plugin feature, we might want to add this before releasing~

The setup process requires azure and Teams access. Currently only @marcoandredinis and I have access to the test env.

### CODE REVIEW

File: examples/resources/plugins/teleport-ms-teams.toml [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+# client_key = "/var/lib/teleport/plugins/ms-teams/auth.key" # Teleport TLS secret key
+# client_crt = "/var/lib/teleport/plugins/ms-teams/auth.crt" # Teleport TLS certificate
+# root_cas = "/var/lib/teleport/plugins/ms-teams/auth.cas"   # Teleport CA certs
+addr = "localhost:3025"
+identity = "identity"
+
+[msapi]
+# MS API ID's. Please, check the documentation.
+app_id = "xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx"
+app_secret = "XXXXXXXXXXXXXXXXXXXXXX"
```
Comment (marcoandredinis) [inline examples/resources/plugins/teleport-ms-teams.toml:28]:
```suggestion
app_secret = "XXXXXXXXXXXXXXXXXXXXXX" # either the app_secret or the absolute path for a file containing the secret
```

**Comment by r0mant** [comment]
@ptgott Can you please review this?

**Comment by ptgott** [comment]
@r0mant I'll make time to do this tomorrow.

**Comment by alexfornuto** [comment]
I suggest using the full name "Microsoft Teams".

File: examples/resources/plugins/teleport-msteams.toml [hunk 2]
```diff
@@ -0,0 +1,49 @@
+# Example ms teams plugin configuration TOML file
```
Comment (ptgott) [inline examples/resources/plugins/teleport-msteams.toml:1]:
```suggestion
# Example Microsoft Teams plugin configuration TOML file
```

File: examples/resources/plugins/teleport-msteams.toml [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+# If true, recipients existence got checked on plugin start
```
Comment (ptgott) [inline examples/resources/plugins/teleport-msteams.toml:3]:
```suggestion
# If true, the recipient's existence is checked on plugin start.
```

File: examples/resources/plugins/teleport-msteams.toml [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+# When a recipient is checked, the app is installed for the user if it was not already.
+# This takes some time and does not fits well with HTTP timeouts, hence preload is necessary when a new
```
Comment (ptgott) [inline examples/resources/plugins/teleport-msteams.toml:5]:
```suggestion
# This takes some time and does not fit well with HTTP timeouts, hence preload is necessary when a new
```

File: examples/resources/plugins/teleport-msteams.toml [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+# recipient is added.
+preload = true
```
Comment (ptgott) [inline examples/resources/plugins/teleport-msteams.toml:7]: I think the comment here needs a bit more context. Maybe move the comment text into a `Details` box where we explain different configuration fields. In particular, I'm wondering:
- "recipient": what does this mean in this context?
- what does "does not fits well with HTTP timeouts" mean? The plugin might fail to start due to recipient checking exceeding HTTP timeouts on the MS Teams API backend?

**Review by ptgott — COMMENTED** [review]
For better SEO, I also think we should also change `msteams` to `microsoft-teams` in the guide URL.

File: docs/pages/access-controls/access-request-plugins/ssh-approval-msteams.mdx [hunk 6]
```diff
@@ -0,0 +1,460 @@
...
+[teleport]
+addr = "mytenant.teleport.sh:443"
+identity = "/var/lib/teleport/plugins/msteams/auth.pem"
+```
+
+</ScopedBlock>
+ 
+**`[role_to_recipients]`**
+
+The `role_to_recipients` map configure the teams users and channels that the
```
[the remark this answers is not in the document]
Comment (hugoShaka) [reply docs/pages/access-controls/access-request-plugins/ssh-approval-msteams.mdx:272]:
```suggestion
The `role_to_recipients` map configure the users, and channels that the
```

File: docs/pages/access-controls/access-request-plugins/ssh-approval-msteams.mdx [hunk 7]
```diff
@@ -0,0 +1,460 @@
...
+compromising productivity.
+
+## Prerequisites
+
+(!/docs/pages/includes/commercial-prereqs-tabs.mdx!)
+
+- A MsTeams License (Microsoft 365 Business)
+- Azure console access in the organization/directory paying the Office 365 Licenses (MsTeams licences)
+- An Azure Resource Group in the same directory, that will host the access-request plugin resources.
+  You should have enough permissions to create and edit Bot services in this Resource Group.
```
[the remark this answers is not in the document]
Comment (hugoShaka) [reply docs/pages/access-controls/access-request-plugins/ssh-approval-msteams.mdx]: I have no idea, cloud permissions can be granted in a lot of different fashions and scopes, the only test to do would be trying to create one

File: docs/pages/access-controls/access-request-plugins/ssh-approval-msteams.mdx [hunk 8]
```diff
@@ -0,0 +1,460 @@
...
+$ sudo mkdir -p /var/lib/teleport/plugins/msteams
+$ sudo mv auth.* /var/lib/teleport/plugins/msteams
+```
+
+## Step 5/9. Register an Azure Bot
+
+The Access Request plugin for MsTeams receives Access Request events from the
+Teleport Auth Service, formats them into MsTeams messages, and sends them to the
+MsTeams API to post them in your workspace. For this to work, you must register a
+new Azure Bot.
```
[the remark this answers is not in the document]
Comment (hugoShaka) [reply docs/pages/access-controls/access-request-plugins/ssh-approval-msteams.mdx:105]: Done

File: docs/pages/access-controls/access-request-plugins/ssh-approval-msteams.mdx [hunk 9]
```diff
@@ -0,0 +1,460 @@
...
+these to the list of channels to notify. If a suggested reviewer is an email
+address, the plugin will look up the the direct message channel for that
+address and post a message in that channel.
+
+</Details>
+
+Configure the MsTeams plugin to notify you when a user requests the `editor` role
+by adding the following to your `role_to_recipients` config (replace
+`TELEPORT_USERNAME` with the email of the user you assigned the `editor-reviewer`
+role earlier, and ):
```
[the remark this answers is not in the document]
Comment (hugoShaka) [reply docs/pages/access-controls/access-request-plugins/ssh-approval-msteams.mdx]: fixed

File: docs/pages/access-controls/access-request-plugins/ssh-approval-msteams.mdx [hunk 10]
```diff
@@ -0,0 +1,460 @@
...
+$ teleport-msteams configure /var/lib/teleport/plugins/msteams/assets --appID <Microsoft App ID value> --tenantID <App Tenant ID value> --appSecret <password value saved on step #4>
+```
+
+This should result in a config file like the one below:
+
+```toml
+(!examples/resources/plugins/teleport-msteams.toml!)
+```
+
+Recover the `/var/lib/teleport/plugins/msteams/assets/app.zip` file,
```
[the remark this answers is not in the document]
Comment (hugoShaka) [reply docs/pages/access-controls/access-request-plugins/ssh-approval-msteams.mdx]: done

File: docs/pages/access-controls/access-request-plugins/ssh-approval-msteams.mdx [hunk 11]
```diff
@@ -0,0 +1,460 @@
...
+The result should be as follows:
+
+![Granted App Permissions](../../../img/enterprise/plugins/msteams/granted-app-permissions.png)
+
+## Step 6/9. Configure the Teleport MsTeams plugin
+
+At this point, the Teleport MsTeams plugin has the credentials it needs to
+communicate with your Teleport cluster and Azure APIs, but the app has not been
+installed to MsTeams yet. In this step, you will configure the MsTeams plugin
+to use the Azure credentials and generate the MsTeams app package. You will
```
[the remark this answers is not in the document]
Comment (hugoShaka) [reply docs/pages/access-controls/access-request-plugins/ssh-approval-msteams.mdx]: Added sentence describing the package purpose

File: docs/pages/access-controls/access-request-plugins/ssh-approval-msteams.mdx [hunk 12]
```diff
@@ -0,0 +1,460 @@
...
+contact an admin user to grant the permissions.
+
+Once permissions have been approved, refresh the page and check the approval status.
+The result should be as follows:
+
+![Granted App Permissions](../../../img/enterprise/plugins/msteams/granted-app-permissions.png)
+
+## Step 6/9. Configure the Teleport MsTeams plugin
+
+At this point, the Teleport MsTeams plugin has the credentials it needs to
```
[the remark this answers is not in the document]
Comment (hugoShaka) [reply docs/pages/access-controls/access-request-plugins/ssh-approval-msteams.mdx:175]: Yes, I rephrased the next sentence to try to make it clearer

File: docs/pages/access-controls/access-request-plugins/ssh-approval-msteams.mdx [hunk 13]
```diff
@@ -0,0 +1,460 @@
...
+- `TeamsAppInstallation.ReadWriteSelfForUser.All` (used to initiate communication with
+   a user that never interacted with the Teams app before)
+- `TeamsAppInstallation.ReadWriteSelfForTeam.All` (used to discover if the app is installed
+   in the Team)
+
+![Specify App Permissions](../../../img/enterprise/plugins/msteams/specify-app-permissions.png)
+
+At this point the app declares the required permissions but those have not been granted.
+
+If you are admin, click "Grant admin consent for \<directory name\>". If you are not admin,
```
[the remark this answers is not in the document]
Comment (hugoShaka) [reply docs/pages/access-controls/access-request-plugins/ssh-approval-msteams.mdx]: done

File: docs/pages/access-controls/access-request-plugins/ssh-approval-msteams.mdx [hunk 14]
```diff
@@ -0,0 +1,460 @@
...
+
+![Manage Bot App](../../../img/enterprise/plugins/msteams/manage-bot-app.png)
+
+Then, go to the "Certificates & Secrets" section, choose to create a "New client secret".
+Use the "Copy" icon to recover the newly created secret and keep it with the
+previously recovered App ID and Tenant ID.
+
+### Specify the permissions used by the app
+
+Still in the app management view ("Configuration", then "Manage" the Microsoft App ID),
```
[the remark this answers is not in the document]
Comment (hugoShaka) [reply docs/pages/access-controls/access-request-plugins/ssh-approval-msteams.mdx:149]: done

File: docs/pages/access-controls/access-request-plugins/ssh-approval-msteams.mdx [hunk 15]
```diff
@@ -0,0 +1,460 @@
...
+
+The client secret will be used by the Teleport plugin to authenticate as the bot's app when
+searching users and posting messages.
+
+On the bot's "Configuration" tab, find "Microsoft App ID", click "Manage" link next to it.
+
+![Manage Bot App](../../../img/enterprise/plugins/msteams/manage-bot-app.png)
+
+Then, go to the "Certificates & Secrets" section, choose to create a "New client secret".
+Use the "Copy" icon to recover the newly created secret and keep it with the
```
[the remark this answers is not in the document]
Comment (hugoShaka) [reply docs/pages/access-controls/access-request-plugins/ssh-approval-msteams.mdx]: done

File: docs/pages/access-controls/access-request-plugins/ssh-approval-msteams.mdx [hunk 16]
```diff
@@ -0,0 +1,460 @@
...
+Once the bot is created, open the resource and go the the "Channels" tab. Click
+"Microsoft Teams" and add the MsTeams channel.
+
+The result should be as follows:
+
+![Add Bot Channel](../../../img/enterprise/plugins/msteams/add-bot-channel.png)
+
+### Recover the bot app ID and the directory tenant ID
+
+On the bot's "Configuration" tab, recover the values of "Microsoft App ID" and "App Tenant ID".
```
[the remark this answers is not in the document]
Comment (hugoShaka) [reply docs/pages/access-controls/access-request-plugins/ssh-approval-msteams.mdx]: Done

File: docs/pages/access-controls/access-request-plugins/ssh-approval-msteams.mdx [hunk 17]
```diff
@@ -0,0 +1,460 @@
...
+Those two UUIDs will be used in the plugin configuration.
+
+### Create a client secret
+
+The client secret will be used by the Teleport plugin to authenticate as the bot's app when
+searching users and posting messages.
+
+On the bot's "Configuration" tab, find "Microsoft App ID", click "Manage" link next to it.
```
[the remark this answers is not in the document]
Comment (hugoShaka) [reply docs/pages/access-controls/access-request-plugins/ssh-approval-msteams.mdx]: Done

File: docs/pages/access-controls/access-request-plugins/ssh-approval-msteams.mdx [hunk 18]
```diff
@@ -0,0 +1,460 @@
...
+The result should be as follows:
+
+![Add Bot Channel](../../../img/enterprise/plugins/msteams/add-bot-channel.png)
+
+### Recover the bot app ID and the directory tenant ID
+
+On the bot's "Configuration" tab, recover the values of "Microsoft App ID" and "App Tenant ID".
+Those two UUIDs will be used in the plugin configuration.
+
+### Create a client secret
```
[the remark this answers is not in the document]
Comment (hugoShaka) [reply docs/pages/access-controls/access-request-plugins/ssh-approval-msteams.mdx]: Done

File: docs/pages/access-controls/access-request-plugins/ssh-approval-msteams.mdx [hunk 19]
```diff
@@ -0,0 +1,460 @@
...
+the resource group and the bot pricing tier.
+
+In the "Microsoft App ID" section choose "Single Tenant" and "Create new
+Microsoft App ID".
+
+![Create Azure Bot](../../../img/enterprise/plugins/msteams/create-azure-bot.png)
+
+### Connect the bot to MsTeams
+
+Once the bot is created, open the resource and go the the "Channels" tab. Click
```
[the remark this answers is not in the document]
Comment (hugoShaka) [reply docs/pages/access-controls/access-request-plugins/ssh-approval-msteams.mdx]: Rephrased

File: docs/pages/access-controls/access-request-plugins/ssh-approval-msteams.mdx (hunk 8 above)
[the remark this answers is not in the document]
Comment (hugoShaka) [reply docs/pages/access-controls/access-request-plugins/ssh-approval-msteams.mdx]: Done

File: docs/pages/access-controls/access-request-plugins/ssh-approval-msteams.mdx [hunk 20]
```diff
@@ -0,0 +1,460 @@
...
+```
+
+## Step 5/9. Register an Azure Bot
+
+The Access Request plugin for MsTeams receives Access Request events from the
+Teleport Auth Service, formats them into MsTeams messages, and sends them to the
+MsTeams API to post them in your workspace. For this to work, you must register a
+new Azure Bot.
+
+### Register a new Azure bot
```
[the remark this answers is not in the document]
Comment (hugoShaka) [reply docs/pages/access-controls/access-request-plugins/ssh-approval-msteams.mdx:108]: Done

File: docs/pages/access-controls/access-request-plugins/ssh-approval-msteams.mdx [hunk 21]
```diff
@@ -0,0 +1,460 @@
...
+(!/docs/pages/includes/commercial-prereqs-tabs.mdx!)
+
+- A MsTeams License (Microsoft 365 Business)
+- Azure console access in the organization/directory paying the Office 365 Licenses (MsTeams licences)
+- An Azure Resource Group in the same directory, that will host the access-request plugin resources.
+  You should have enough permissions to create and edit Bot services in this Resource Group.
+- Someone with Global Admin rights on the Azure Active Directory that will approve
+  the bot application permissions
+- Someone with the `Teams administrator` role that will approve the MsTeams
+  application installation request
```
[the remark this answers is not in the document]
Comment (hugoShaka) [reply docs/pages/access-controls/access-request-plugins/ssh-approval-msteams.mdx]: done

File: docs/pages/access-controls/access-request-plugins/ssh-approval-msteams.mdx [hunk 22]
```diff
@@ -0,0 +1,460 @@
...
+The Teleport MsTeams plugin uses a config file in TOML format.
+The `configure` subcommand generates a directory containing the TOML
+configuration file and a `app.zip` file that will be used later to
+add the MsTeams application into the organization catalog.
+
+The command will create the directory `/var/lib/teleport/plugins/msteams/assets`
+(might require root/sudo):
+
+```code
+$ teleport-msteams configure /var/lib/teleport/plugins/msteams/assets --appID <Microsoft App ID value> --tenantID <App Tenant ID value> --appSecret <password value saved on step #4>
```
[the remark this answers is not in the document]
Comment (hugoShaka) [reply docs/pages/access-controls/access-request-plugins/ssh-approval-msteams.mdx]: Done

File: docs/pages/access-controls/access-request-plugins/ssh-approval-msteams.mdx [hunk 23]
```diff
@@ -0,0 +1,460 @@
...
+previously recovered App ID and Tenant ID.
+
+### Specify the permissions used by the app
+
+Still in the app management view ("Configuration", then "Manage" the Microsoft App ID),
+go to the "API permissions" tab.
+
+Add the following Microsoft Graph Application permissions:
+
+- `AppCatalog.Read.All` (used to list Teams app and check the app is installed)
```
[the remark this answers is not in the document]
Comment (hugoShaka) [reply docs/pages/access-controls/access-request-plugins/ssh-approval-msteams.mdx]: Done

**Comment by r0mant** [comment]
@ptgott Could you please take another look? Would be nice to release these docs as the MS Teams plugin is already published.

File: docs/pages/access-controls/access-request-plugins/ssh-approval-msteams.mdx [hunk 24]
```diff
@@ -0,0 +1,463 @@
...
+## Prerequisites
+
+(!/docs/pages/includes/commercial-prereqs-tabs.mdx!)
+
+- A Microsoft Teams License (Microsoft 365 Business)
+- Azure console access in the organization/directory paying the Office 365 Licenses (MsTeams licences)
+- An Azure resource group in the same directory. This will  host resources for the the Microsoft Teams Access Request plugin.
+  You should have enough permissions to create and edit Azure Bot Services in this resource group.
+- Someone with Global Admin rights on the Azure Active Directory that will
+  grant permissions to the plugin
```
Comment (ptgott) [inline docs/pages/access-controls/access-request-plugins/ssh-approval-msteams.mdx:21]:
```suggestion
  grant permissions to the plugin.
```

File: docs/pages/access-controls/access-request-plugins/ssh-approval-msteams.mdx [hunk 25]
```diff
@@ -0,0 +1,460 @@
...
+
+The Access Request plugin for MsTeams receives Access Request events from the
+Teleport Auth Service, formats them into MsTeams messages, and sends them to the
+MsTeams API to post them in your workspace. For this to work, you must register a
+new Azure Bot.
+
+### Register a new Azure bot
+
+Visit [https://portal.azure.com/#create/Microsoft.AzureBot](https://portal.azure.com/#create/Microsoft.AzureBot)
+to create a new bot. Choose the bot handle (this is for semantic use only, it will
```
[the remark this answers is not in the document]
Comment (ptgott) [reply docs/pages/access-controls/access-request-plugins/ssh-approval-msteams.mdx:111]:
```suggestion
to create a new bot. Choose the bot handle so you can find the bot later in the Azure console (the bot handle will
```

File: docs/pages/access-controls/access-request-plugins/ssh-approval-msteams.mdx [hunk 26]
```diff
@@ -0,0 +1,463 @@
...
+MsTeams API to post them in your workspace. For this to work, you must register a
+new Azure Bot. Azure Bot is a managed service by Microsoft that allows to
+develop bots that interact with users through different channels, including
+Microsoft Teams.
+
+### Register a new Azure bot
+
+Visit [https://portal.azure.com/#create/Microsoft.AzureBot](https://portal.azure.com/#create/Microsoft.AzureBot)
+to create a new bot. Choose the bot handle (this is for semantic use only, it will
+not be displayed to the user or used in configuration), the Azure subscription,
```
Comment (ptgott) [inline docs/pages/access-controls/access-request-plugins/ssh-approval-msteams.mdx:112]:
```suggestion
not be displayed to the user or used to configure the Microsoft Teams plugin). Also edit the Azure subscription,
```

File: docs/pages/access-controls/access-request-plugins/ssh-approval-msteams.mdx [hunk 27]
```diff
@@ -0,0 +1,463 @@
...
+the resource group and the bot pricing tier.
+
+In the "Microsoft App ID" section choose "Single Tenant" and "Create new
+Microsoft App ID".
+
+![Create Azure Bot](../../../img/enterprise/plugins/msteams/create-azure-bot.png)
+
+### Connect the bot to MsTeams
+
+Once the bot is created, open its resource page on the azure console and
```
Comment (ptgott) [inline docs/pages/access-controls/access-request-plugins/ssh-approval-msteams.mdx:122]:
```suggestion
Once the bot is created, open its resource page on the Azure console and
```

File: docs/pages/access-controls/access-request-plugins/ssh-approval-msteams.mdx [hunk 28]
```diff
@@ -0,0 +1,463 @@
...
+navigate to the "Channels" tab. Click "Microsoft Teams" and add the MsTeams
+channel.
+
+The result should be as follows:
+
+![Add Bot Channel](../../../img/enterprise/plugins/msteams/add-bot-channel.png)
+
+### Recover the bot app ID, tenant ID and create a client secret
```
Comment (ptgott) [inline docs/pages/access-controls/access-request-plugins/ssh-approval-msteams.mdx:130]:
```suggestion
### Obtain information about your Microsoft App
```
Suggesting a more general heading

File: docs/pages/access-controls/access-request-plugins/ssh-approval-msteams.mdx [hunk 29]
```diff
@@ -0,0 +1,463 @@
...
+
+## Step 6/9. Configure the Teleport MsTeams plugin
+
+At this point, the Teleport MsTeams plugin has the credentials it needs to
+communicate with your Teleport cluster and Azure APIs, but the app has not been
+installed to MsTeams yet.
+
+In this step, you will configure the MsTeams plugin to use the Azure
+credentials and generate the Teams App package that will be used to install the
+Teams App to your teams instance. You will also configure the plugin to notify
```
Comment (ptgott) [inline docs/pages/access-controls/access-request-plugins/ssh-approval-msteams.mdx:181]:
```suggestion
Microsoft Teams App. You will also configure the plugin to notify
```

File: docs/pages/access-controls/access-request-plugins/ssh-approval-msteams.mdx [hunk 30]
```diff
@@ -0,0 +1,463 @@
...
+```
+
+This should result in a config file like the one below:
+
+```toml
+(!examples/resources/plugins/teleport-msteams.toml!)
+```
+
+Copy the `/var/lib/teleport/plugins/msteams/assets/app.zip` file to your local
+computer, you will have to upload it to MsTeams later.
```
Comment (ptgott) [inline docs/pages/access-controls/access-request-plugins/ssh-approval-msteams.mdx:205]:
```suggestion
computer. You will have to upload it to Microsoft Teams later.
```

**Review by ptgott — APPROVED** [review]
Approved with a list of minor suggestions. The main one is to replace `MsTeams` with `Microsoft Teams` in all occurrences.

File: docs/pages/access-controls/access-request-plugins/ssh-approval-msteams.mdx [hunk 31]
```diff
@@ -0,0 +1,463 @@
+---
+title: Access Requests with Microsoft Teams
+description: How to set up Teleport's MsTeams plugin for privilege elevation approvals.
+---
+
+This guide will explain how to set up MsTeams to receive Access Request messages
```
[the remark this answers is not in the document]
Comment (hugoShaka) [reply docs/pages/access-controls/access-request-plugins/ssh-approval-msteams.mdx]: Done

**Review by r0mant — APPROVED** [review]

**Pull request merged**